DM-14821: mechanisms for packing (some) data IDs into integers

TallJimbo opened the pull request.

TallJimbo committed:
Add maximums for dependent dimensions to Instrument table.

These are needed for packing Visit+Detector and Exposure+Detector
data IDs into integers.  Putting them in the Registry database saves
us from having to import camera-specific Python code just to compute
those packed IDs.

TallJimbo committed:
Doc and whitespace cleanups.

TallJimbo committed:
Add custom dictionary class with DimensionElement keys.

TallJimbo committed:
Add pack-to-int-or-blob mechanism for data IDs.

TallJimbo committed:
Add missing __init__.py lift-import.

TallJimbo committed:
Add flexibility to DataId construction.

It's now possible to expand the dimensions of an existing DataId
by passing extra link values as kwargs or via the "extra" arg.

We also support passing DimensionSet as the dimensions argument.

TallJimbo committed:
Add metadata columns to SkyMap table.

TallJimbo committed:
Add SkyMap DataIdPackers to Registry config.

TallJimbo committed:
Implement implied flexibility in DataId construction.

Passing **kwds with new link values along with an existing DataId
now creates a new DataId with those link values changed, rather
than ignoring those kwds.

Unfortunately, this is the latest in a sequence of changes that have
together made DataId construction and expansion *quite* complicated
and hard to follow.  I don't think a refactor is appropriate for this
ticket (it's at least part of DM-17023), but it's now definitely
overdue.

TallJimbo committed:
Just pass full registry when constructing SqlPreFlight helper.

TallJimbo committed:
Expand DataIds to include packers in selectDimensions.

natelust commented on config/schema.yaml:
```diff
@@ -450,6 +450,24 @@ schema:
+          this instrument (exclusive).
+      -
+        name: exposure_max
+        type: int
```
How are you sure there is a max like this (and above)? Is it based on the fact that some instruments use 16,32,64,etc bit numbers for identifiers? If so what if an instrument uses say 128, and python int is 64? Or is this int actually a hint to sql and nothing more, and they just have a "number" field. What if a telescope does not use numbers to increment, like some sort of lexographic numbering based on some datestring?

timj replied:
The metadata translator constructs the exposure_id as an integer regardless of how the instrument is defining it in the headers. Python integers can be any size.

TallJimbo replied:
I'm basically reasoning from the other direction: I need an `exposure_max` or something equivalent to implement the ID packers we need to support (Gen2 uses a number of bits, which I think just obfuscates the fact that it defines a maximum exposure number).  So I'm essentially putting it on the `Instrument` subclass (and hence the obs package) to define something and stick to it.  Given the size of even just an int32 and the number of exposures that can practically taken with real instruments of the sort we're trying to support, I don't think that should be a problem.

natelust commented on python/lsst/daf/butler/core/dimensions/dataId.py:
```diff
@@ -19,12 +19,155 @@
+            self._keys = DimensionSet(universe, byName.keys())
+
+        if where is None:
+            where = lambda x: True  # noqa
```
Maybe use _ in the lambda to emphasize that this is never going to be used

timj replied:
I thought we frowned upon lambda in this context and preferred:
```python
def where():
    return True
```
?

timj replied:
PS `noqa` is meant to include the code that you are ignoring.

TallJimbo replied:
Agree with `_`.

I generally disagree with PEP8 on local `def` being better than assigning a lambda to a variable (I don't know of any other language with that preference).  Obviously it's a bad idea to pick a fight with PEP8 in general, but relying on a local `def` to shadow a kwarg seemed like a particularly bad idea.  I'll certainly look up the code for `noqa`, though.

natelust commented on python/lsst/daf/butler/core/dimensions/dataId.py:
```diff
@@ -19,12 +19,155 @@
+        except AttributeError:
+            return NotImplemented
+
+    def updateValues(self, other):
```
So I think it would be good to talk about interior v exterior mutability at some point. If someone reads the documentation for the class, they may take this class as "immutable" but this method implies that it is mutable (though I know it is just interior mutability) however these are not normally python concepts, and someone might be surprised by calling items, getting a list of the items, and then later find that that list does not match what is in the DimensionsKeyDict. I dont' know that we want to disallow this behavior, but it can be surprising, so we should think if this is the ideal structure, or what kind of documentation on the promises this class makes needs to be written.

TallJimbo replied:
I had written a brief blurb on this in the `Notes` section of the class doc; I'll expand that.  It does turn out to be a very convenient pattern for the dimension system, in which there are frequently contexts in which the set of dimensions is fixed, and we already have an immutable `DimensionSet` class to use as the return value for `keys()`.

natelust commented on python/lsst/daf/butler/core/registry.py:
```diff
@@ -888,13 +890,11 @@ def expandDataId(self, dataId=None, *, dimension=None, metadata=None, region=Fal
                 # If a single dimension was passed explicitly, permit
                 # 'metadata' to be a sequence corresponding to just that
                 # dimension (by normalizing it into a dict-of-sets now).
```
I think this comment needs updated

natelust commented on python/lsst/daf/butler/core/dataIdPacker.py:
```diff
@@ -0,0 +1,464 @@
+    """
+
+    # Override __new__ so subclasses aren't forced to call super().__init__
+    # and have to support at least the standard signature.
```
This message could be clearer

natelust commented on python/lsst/daf/butler/core/dataIdPacker.py:
```diff
@@ -0,0 +1,464 @@
+        matched = []
+        for given, loaders in self._loaders.items():
+            if given.issubset(dimensions):
+                for name, loader in loaders.items():
```
it is confusing to have loaders be part of a tuple on items in the dictionary named _loaders

natelust reviewed (changes requested):
We talked about moving the packer object lookups onto the registry, and have two different calls to either pick if the given dataId is already equal to the required, and one to just get the packer.

I would like to see that work before this merges.

TallJimbo committed:
Improve docs and comments in DimensionKeyDict.

TallJimbo committed:
Fix code comment in expandDataId.

TallJimbo committed:
Simplify DataIdPacker mechanisms.

We now put the need metadata fields in the Registry config, so we
don't need to import anything in order to know how to expand Data IDs.
That means packer classes are now truly only imported on first use.

As work on DM-16227 eliminated the previous use case for packers with
given | covered != required, I've also removed "covered" since it's
now redundant.

Finally, packers are no longer attached to DataIds, and are instead
obtained directly from Registry.

TallJimbo commented:
@natelust, I've made the changes we discussed, and all of the line comments I didn't respond to have become obsolete as a result.  Please take another look when you get a chance.
